### PR TITLE
Add VSTypeScriptEditorInlineRenameService + associated classes

### PR DESCRIPTION
**Pull request by allisonchou**
Related to https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1077984

Declares new property "DefinitionLocations".

### CODE REVIEW

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptInlineRenameInfoBase.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        public abstract bool ForceRenameOverloads { get; }
+        public abstract string DisplayName { get; }
+        public abstract string FullDisplayName { get; }
+        public abstract Glyph Glyph { get; }
+        public abstract ImmutableArray<DocumentSpan> DefinitionLocations { get; }
+
+        public abstract Task<IInlineRenameLocationSet> FindRenameLocationsAsync(OptionSet optionSet, CancellationToken cancellationToken);
+        public abstract TextSpan? GetConflictEditSpan(InlineRenameLocation location, string replacementText, CancellationToken cancellationToken);
+        public abstract string GetFinalSymbolName(string replacementText);
+        public abstract TextSpan GetReferenceEditSpan(InlineRenameLocation location, CancellationToken cancellationToken);
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptInlineRenameInfoBase.cs]: We need to replace internal types like `Glyph`, `InlineRenameLocation` and also class for `IInlineRenameLocationSet`. 

Turns out F# already does all this:
http://sourceroslyn.io/#Microsoft.CodeAnalysis.ExternalAccess.FSharp/Internal/Editor/FSharpEditorInlineRenameService.cs,106

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptGlpyhHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+using System;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
+{
+    internal static class VSTypeScriptGlyphHelpers
+    {
+        public static Microsoft.CodeAnalysis.Glyph ConvertTo(VSTypeScriptGlyph glyph)
+        {
+            switch (glyph)
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptGlpyhHelpers.cs]: Do we show "convert to switch expression" suggestion? If so, please make the change.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
+{
+    internal static class VSTypeScriptInlineRenameReplacementKindHelpers
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs]: >internal [](start = 4, length = 8)

Let's split this file into multiple - one file per class.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    {
+        public static InlineRenameReplacementKind ConvertTo(VSTypeScriptInlineRenameReplacementKind kind)
+        {
+            switch (kind)
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs]: >switch [](start = 12, length = 6)

Switch expression

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            }
+        }
+    }
+
+    internal class VSTypeScriptInlineRenameInfo : IInlineRenameInfo
+    {
+        private readonly IVSTypeScriptInlineRenameInfo _info;
+
+        public VSTypeScriptInlineRenameInfo(IVSTypeScriptInlineRenameInfo info)
+        {
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs]: >{ [](start = 8, length = 1)

Add null check - `Contract.ThrowIfNull(info)`

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    internal class VSTypeScriptInlineRenameLocationSet : IInlineRenameLocationSet
+    {
+        private readonly IVSTypeScriptInlineRenameLocationSet _set;
+        private readonly IList<InlineRenameLocation> _locations;
+
+        public VSTypeScriptInlineRenameLocationSet(IVSTypeScriptInlineRenameLocationSet set)
+        {
+            _set = set;
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs]: >_set [](start = 12, length = 4)

Add null check - `Contract.ThrowIfNull(info)`

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        }
+    }
+
+    internal class VSTypeScriptInlineRenameReplacementInfo : IInlineRenameReplacementInfo
+    {
+        private readonly IVSTypeScriptInlineRenameReplacementInfo _info;
+
+        public VSTypeScriptInlineRenameReplacementInfo(IVSTypeScriptInlineRenameReplacementInfo info)
+        {
+            _info = info;
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs]: >_info [](start = 12, length = 5)

Add null check - `Contract.ThrowIfNull(info)`

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+                default:
+                    {
+                        throw ExceptionUtilities.UnexpectedValue(kind);
+                    }
+            }
+        }
+    }
+
+    internal class VSTypeScriptInlineRenameReplacementInfo : IInlineRenameReplacementInfo
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs]: >internal [](start = 4, length = 8)

Make these classes `sealed` - we don't expect anyone to derive from them.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    internal enum VSTypeScriptInlineRenameReplacementKind
+    {
+        NoConflict,
+        ResolvedReferenceConflict,
+        ResolvedNonReferenceConflict,
+        UnresolvedConflict,
+        Complexified,
+    }
+
+    internal readonly struct VSTypeScriptInlineRenameReplacement
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs]: >VSTypeScriptInlineRenameReplacement [](start = 29, length = 35)

It seems F# is not following the patterns we have elsewhere. For simple wrappers we have been using pattern like so:

```C#
    internal readonly struct VSTypeScriptInlineRenameReplacementWrapper
    {
        private readonly InlineRenameReplacement _underlyingObject;

        public VSTypeScriptInlineRenameReplacementWrapper(InlineRenameReplacement underlyingObject)
        {
           _underlyingObject = underlyingObject;
        }

        public VSTypeScriptInlineRenameReplacementKind Kind => _underlyingObject.Kind;
        public TextSpan OriginalSpan => _underlyingObject.OriginalSpan;
        public TextSpan NewSpan => _underlyingObject.NewSpan;
    }
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
+{
+    internal readonly struct VSTypeScriptInlineRenameLocation
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs]: >internal [](start = 4, length = 8)

We should use -Wrapper pattern for this one (see my other comment on VSTypeScriptInlineRenameReplacementWrapper)

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptGlpyhHelpers.cs [hunk 11]
```diff
@@ -0,0 +1,327 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
+{
+    internal static class VSTypeScriptGlyphHelpers
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptGlpyhHelpers.cs]: >VSTypeScriptGlyphHelpers [](start = 26, length = 24)

Typo in the file name.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptGlpyhHelpers.cs (hunk 2 above)
Comment (allisonchou) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptGlpyhHelpers.cs]: Surprisingly, there was no suggestion, but I implemented the change anyways.

File: src/Compilers/Core/Portable/Symbols/LanguageNames.cs [hunk 12]
```diff
@@ -28,5 +28,13 @@ public static class LanguageNames
...
         public const string FSharp = "F#";
+
+        /// <summary>
+        /// The common name used for the TypeScript language.
+        /// </summary>
+        /// <remarks>
+        /// TypeScript is not a supported compile target for the Roslyn compiler.
+        /// </remarks>
+        public const string TypeScript = "TypeScript";
     }
```
Comment (tmat) [inline src/Compilers/Core/Portable/Symbols/LanguageNames.cs]: Don't add this one. Roslyn does not implement TypeScript language, so it shouldn't be in the public API. Use `InternalLanguageNames.TypeScript` instead.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
+{
+    internal readonly struct VSTypeScriptInlineRenameLocationWrapper
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs]: >internal [](start = 4, length = 8)

I'd be good to split this file into one file per type as well.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs]: >VSTypeScript [](start = 48, length = 12)

Another thing to think about is what types are going to be actually used by TypeScript. All those types need to go to `Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api` subnamespace (and subfolder).
I think most of these types should be in that category, except for VSTypeScriptEditorInlineRenameService, VSTypeScriptGlyphHelpers, and VSTypeScriptInlineRenameReplacementKindHelpers.

Comment (allisonchou) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/IVSTypeScriptEditorInlineRenameService.cs]: Ah got it, thank you. Just moved the relevant files over the API folder. 👍

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptInlineRenameReplacementInfo.cs [hunk 15]
```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptInlineRenameReplacementInfo.cs]: >VSTypeScript [](start = 48, length = 12)

Missing .Api

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+{
+    [Shared]
+    [ExportLanguageService(typeof(IEditorInlineRenameService), InternalLanguageNames.TypeScript)]
+    internal sealed class VSTypeScriptEditorInlineRenameService : IEditorInlineRenameService
+    {
+        private readonly IVSTypeScriptEditorInlineRenameService _service;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VSTypeScriptEditorInlineRenameService(IVSTypeScriptEditorInlineRenameService service)
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs]: >IVSTypeScriptEditorInlineRenameService [](start = 53, length = 38)

Use `Lazy<IVSTypeScriptEditorInlineRenameService>`

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+using System;
+using Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
+{
+    [Shared]
+    [ExportLanguageService(typeof(IEditorInlineRenameService), InternalLanguageNames.TypeScript)]
+    internal sealed class VSTypeScriptEditorInlineRenameService : IEditorInlineRenameService
+    {
+        private readonly IVSTypeScriptEditorInlineRenameService _service;
```
Comment (tmat) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptEditorInlineRenameService.cs]: >IVSTypeScriptEditorInlineRenameService [](start = 25, length = 38)

 `Lazy<IVSTypeScriptEditorInlineRenameService>`

**Pull request merged**